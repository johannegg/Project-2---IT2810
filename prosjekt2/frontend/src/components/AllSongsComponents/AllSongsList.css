.songContainer {
	margin-left: 20px;
	margin-right: 20px;
	padding: 20px;
	border-radius: 10px;
	box-shadow: 0 4px 8px 0 #00000033;
	width: 100%;
	background-image: linear-gradient(to right, var(--theme-dark-green), Var(--theme-green));
}
.songTable {
	width: 100%;
	border-collapse: collapse;
	margin-top: 20px;
}

.tableRow {
	border-bottom: 1px solid var(--theme-darker-green);
	text-align: left;
	font-size: medium;
}

.tableRow:hover {
	background-color: var(--theme-darker-green);
}

.titleCell {
	font-weight: 500;
}

.viewsCell {
	font-size: small;
}

td {
	padding-left: 5px;
	padding-right: 5px;
}
