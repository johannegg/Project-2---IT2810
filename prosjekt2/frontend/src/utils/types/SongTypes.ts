export type Artist = {
    id: string;
    name: string;
};

export type Genre = {
    name: string;
};

export type SongData = {
    id: string;
    title: string;
    views: number;
    year: number;
    artist: Artist;
    genre: Genre;
    lyrics: string;
};