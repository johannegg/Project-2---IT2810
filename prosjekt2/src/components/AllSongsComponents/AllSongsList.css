.songContainer {
  margin-top: 20px;
  margin-left: 20px;
  margin-right: 20px;
  padding: 20px;
  border-radius: 5px;
  background-color: var(--theme-green);
  box-shadow: 0 4px 8px 0 #00000033;
  width: 100%;
}
.songTable {
	width: 100%;
	border-collapse: collapse;
	margin-top: 20px;
}

.tableRow {
  border-bottom: 1px solid var(--theme-dark-green);
  text-align: left;
  font-size: medium;
}

.tableRow:hover {
  background-color: var(--theme-dark-green);
}

.viewsCell {
  font-size: small;
}
