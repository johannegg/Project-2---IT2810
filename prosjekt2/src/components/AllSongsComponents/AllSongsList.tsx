import { useEffect, useState } from "react";
import { FaEye } from "react-icons/fa";
import { fetchSongs, Song } from "../../utils/FetchMockData";
import { formatViews } from "../../utils/FormatViews";
import "./AllSongsList.css";

export function AllSongsList() {
	const [songs, setSongs] = useState<Song[]>([]);
	const [error, setError] = useState<string | null>(null);
	const [loading, setLoading] = useState<boolean>(true);

	useEffect(() => {
		const loadData = async () => {
			try {
				const data = await fetchSongs();
				setSongs(data);
			} catch {
				setError("Failed to load data");
			} finally {
				setLoading(false);
			}
		};

		loadData();
	}, []);


	if (loading) return <p>Loading songs...</p>;
	if (error) return <p>{error}</p>;

	return (
		<section className="songContainer">
			<table className="songTable">
				{songs.map((song) => (
					// TODO: Add link to each lyrics pace
					<tr key={song.id} className="tableRow">
						<td>{song.title}</td>
						<td>{song.artist}</td>
						<td>{song.year}</td>
						<td className="viewsCell">
							<FaEye style={{ marginRight: "5px" }} />
							{formatViews(song.views)}
						</td>
					</tr>
				))}
			</table>
		</section>
	);
}
