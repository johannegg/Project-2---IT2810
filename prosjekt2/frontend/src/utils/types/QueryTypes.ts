import { SongData } from "./SongTypes";

// Define types for the GraphQL response
export type SongsQueryResponse = {
    songs: SongData[];
};