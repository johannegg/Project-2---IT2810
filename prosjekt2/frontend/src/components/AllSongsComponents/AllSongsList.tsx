import { FaEye } from "react-icons/fa";
import { formatViews } from "../../utils/FormatViews";
import "./AllSongsList.css";
import { useNavigate } from "react-router-dom";
import FavoriteButton from "../FavoriteButton/FavoriteButton";
import { routeChange } from "../../utils/SongRouteChange";
import { SongData } from "../../utils/types/SongTypes";
import PlusMinusButton from "../PlusMinusButton/PlusMinusButton";

type AllSongsListProps = {
	songs: SongData[];
	selectedGenres: string[];
	maxViews: number;
	minViews: number;
	isInPlaylist?: boolean;
	playlistId?: string;
	onSongRemoved?: (songId: string) => void;
};

export function AllSongsList({
	songs,
	isInPlaylist,
	playlistId,
	selectedGenres,
	maxViews,
	minViews,
	onSongRemoved,
}: AllSongsListProps) {
	const navigate = useNavigate();

	// Filter songs based on selected genres and view range
	const filteredSongs = songs
		.filter((song) => (selectedGenres.length > 0 ? selectedGenres.includes(song.genre.name) : true))
		.filter((song) => song.views >= minViews && song.views <= maxViews);

	return (
		<section className="songContainer" aria-label="Songs list">
			{filteredSongs.length === 0 ? (
				<p aria-label="No songs found">No songs found</p>
			) : (
				<table className="songTable" aria-label="Filtered songs table">
					<tbody>
						{filteredSongs.map((song) => (
							<tr
								key={song.id}
								className="tableRow"
								aria-label={`Song: ${song.title} by ${song.artist.name}`}
								// Navigate to song details page unless a button is clicked
								onClick={(e) => {
									if (!(e.target instanceof HTMLButtonElement)) {
										routeChange(song, navigate);
									}
								}}
								tabIndex={0}
								// Handle keyboard navigation for accessibility
								onKeyDown={(e) => {
									if (
										!(e.target instanceof HTMLButtonElement) &&
										(e.key === "Enter" || e.key === " ")
									) {
										e.preventDefault();
										routeChange(song, navigate);
									}
								}}
							>
								<td className="title-artist-cell" aria-label="Song title and artist">
									<span className="titleCell" aria-label={`Title: ${song.title}`}>
										{song.title}
									</span>
									<span className="artistCell" aria-label={`Artist: ${song.artist.name}`}>
										{song.artist.name}
									</span>
								</td>
								<td aria-label={`Year: ${song.year}`}>{song.year}</td>
								<td className="viewsCell" aria-label={`Views: ${song.views}`}>
									{/* Display view count with icon */}
									<FaEye style={{ marginRight: "5px" }} />
									{formatViews(song.views)}
								</td>
								<td className="plusMinusCell">
									<PlusMinusButton
										song={song}
										isInPlaylist={isInPlaylist}
										playlistId={playlistId}
										onSongRemoved={onSongRemoved}
									/>
								</td>
								<td>
									<FavoriteButton song={song} />
								</td>
							</tr>
						))}
					</tbody>
				</table>
			)}
		</section>
	);
}
