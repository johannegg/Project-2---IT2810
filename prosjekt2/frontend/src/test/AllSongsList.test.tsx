import { describe, it, expect, vi } from "vitest";
import { render, fireEvent, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { AllSongsList } from "../components/AllSongsComponents/AllSongsList";
import { SongData } from "../utils/types/SongTypes";

vi.mock("../components/FavoriteButton/FavoriteButton", () => ({
	default: ({ song }: { song: SongData }) => (
		<button data-testid={`favorite-button-${song.id}`}>Favorite</button>
	),
}));

vi.mock("../components/PlusMinusButton/PlusMinusButton", () => ({
	default: ({
		song,
		onSongRemoved,
	}: {
		song: SongData;
		onSongRemoved?: (songId: string) => void;
	}) => (
		<button data-testid={`plus-minus-button-${song.id}`} onClick={() => onSongRemoved?.(song.id)}>
			{onSongRemoved ? "Remove" : "Add"}
		</button>
	),
}));

const mockSongs: SongData[] = [
	{
		id: "1",
		title: "Song One",
		views: 5000,
		year: 2020,
		artist: { id: "a1", name: "Artist One" },
		genre: { name: "pop" },
		lyrics: "Lyrics of song one",
	},
	{
		id: "2",
		title: "Song Two",
		views: 2000,
		year: 2021,
		artist: { id: "a2", name: "Artist Two" },
		genre: { name: "rock" },
		lyrics: "Lyrics of song two",
	},
];

describe("AllSongsList Component", () => {
	const renderWithRouter = (component: React.ReactNode) => {
		return render(<MemoryRouter>{component}</MemoryRouter>);
	};

	it("renders correctly and matches snapshot", () => {
		const { asFragment } = renderWithRouter(
			<AllSongsList songs={mockSongs} selectedGenres={[]} maxViews={10000} minViews={0} />,
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it("filters songs by selectedGenres and maxViews", () => {
		renderWithRouter(
			<AllSongsList songs={mockSongs} selectedGenres={["pop"]} maxViews={6000} minViews={0} />,
		);

		expect(screen.getByText("Song One")).toBeInTheDocument();
		expect(screen.queryByText("Song Two")).not.toBeInTheDocument();
	});

	it("displays 'No songs found' when no songs match filters", () => {
		renderWithRouter(
			<AllSongsList
				songs={mockSongs}
				selectedGenres={["classical"]}
				maxViews={1000}
				minViews={0}
			/>,
		);

		expect(screen.getByText("No songs found")).toBeInTheDocument();
	});

	it("calls onSongRemoved when remove button is clicked", () => {
		const mockOnSongRemoved = vi.fn();

		renderWithRouter(
			<AllSongsList
				songs={mockSongs}
				selectedGenres={[]}
				maxViews={10000}
				minViews={0}
				isInPlaylist
				playlistId="playlist1"
				onSongRemoved={mockOnSongRemoved}
			/>,
		);

		const removeButton = screen.getByTestId("plus-minus-button-1");
		fireEvent.click(removeButton);

		expect(mockOnSongRemoved).toHaveBeenCalledWith("1");
	});

	it("renders FavoriteButton for each song", () => {
		renderWithRouter(
			<AllSongsList songs={mockSongs} selectedGenres={[]} maxViews={10000} minViews={0} />,
		);

		mockSongs.forEach((song) => {
			expect(screen.getByTestId(`favorite-button-${song.id}`)).toBeInTheDocument();
		});
	});
});
