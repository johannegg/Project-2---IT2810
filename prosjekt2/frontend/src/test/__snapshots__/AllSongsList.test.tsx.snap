// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AllSongsList Component > renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="songContainer"
  >
    <table
      class="songTable"
    >
      <tbody>
        <tr
          class="tableRow"
          tabindex="0"
        >
          <td
            class="title-artist-cell"
          >
            <span
              class="titleCell"
            >
              Song One
            </span>
            <span
              class="artistCell"
            >
              Artist One
            </span>
          </td>
          <td>
            2020
          </td>
          <td
            class="viewsCell"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="margin-right: 5px;"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
              />
            </svg>
            5K
          </td>
          <td
            class="plusMinusCell"
          >
            <button
              data-testid="plus-minus-button-1"
            >
              Add
            </button>
          </td>
          <td>
            <button
              data-testid="favorite-button-1"
            >
              Favorite
            </button>
          </td>
        </tr>
        <tr
          class="tableRow"
          tabindex="0"
        >
          <td
            class="title-artist-cell"
          >
            <span
              class="titleCell"
            >
              Song Two
            </span>
            <span
              class="artistCell"
            >
              Artist Two
            </span>
          </td>
          <td>
            2021
          </td>
          <td
            class="viewsCell"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="margin-right: 5px;"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
              />
            </svg>
            2K
          </td>
          <td
            class="plusMinusCell"
          >
            <button
              data-testid="plus-minus-button-2"
            >
              Add
            </button>
          </td>
          <td>
            <button
              data-testid="favorite-button-2"
            >
              Favorite
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</DocumentFragment>
`;
